### ISSUE TITLE
Could compiler-generated methods like readThis, writeThis, hash be implemented in module code?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#19101]
Today, if the user has not provided certain methods on their types—for example, readThis, writeThis, and hash—the compiler will provide defaults for them in most or all cases (see https://github.com/chapel-lang/chapel/issues/18577 for some discussion on "most" cases for hash()).  In this issue, I'm curious whether the language and Reflection module is currently sufficiently powerful enough to implement these methods as module code rather than in the compiler.  This is something I was curious about yesterday and that Andrew also touched upon today [in a comment](https://github.com/chapel-lang/chapel/issues/19074#issuecomment-1021499043).

**Comment by bradcray** [comment]
A related question is: If it is, or were, powerful enough, would this change anything about our discussions on what/how to name these methods (e.g., https://github.com/chapel-lang/chapel/issues/19038, https://github.com/chapel-lang/chapel/issues/18340)?  My intuition is "probably not" because the user would still need to have a name to use when defining their own method even if we didn't require the compiler to be aware of that name, but I thought it was worth asking in case someone sees something I didn't.

**Comment by dlongnecke-cray** [comment]
We already have `readThisDefaultImpl` and `writeThisDefaultImpl` in `modules/standard/ChapelIO.chpl` which uses `Reflection` to write the contents of class and record types. AFAIK all the default generated function does is call this and pass the concrete type to instantiate it. I think it would be trivial to also do this for `hash` as well.

**Comment by mppf** [comment]
One issue is that I don't think we'd want a Reflection query to find a tertiary method for one of these. Rather, if it does not exist at the record/class definition module, it should be defined by the compiler / use the default implementation.

**Comment by bradcray** [comment]
> One issue is that I don't think we'd want a Reflection query to find a tertiary method for one of these.

For our current uses, it couldn't, though, could it?  E.g., if `IO.chpl` used reflection to look for a `writeThis()` and work around the lack of one if it didn't using a fielderator, it wouldn't find tertiary methods because it wouldn't `use`/`import` the module containing that tertiary method because it doesn't know about it.  Similar for `ChapelHashtable.chpl` and `hash()`.  Am I missing something?

I suppose if a user wrote their own library that relied on writeThis() or hash() that they could use a module that defined a tertiary method and therefore get a different behavior than we did in our standard modules.  But that's on them for having used/imported the module defining the tertiary method isn't it?

**Comment by mppf** [comment]
> For our current uses, it couldn't, though, could it?  E.g., if `IO.chpl` used reflection to look for a `writeThis()` and work around the lack of one if it didn't using a fielderator, it wouldn't find tertiary methods because it wouldn't `use`/`import` the module containing that tertiary method because it doesn't know about it.  Similar for `ChapelHashtable.chpl` and `hash()`.  Am I missing something?

Yes, IO.chpl will find these through the point-of-instantiation rule. Say a user in their own module defines `record R`. This is not visible to IO.chpl at all other than through point-of-instantiation. The point-of-instantiation can include tertiary methods in addition to or instead of the ones in the module.

**Comment by bradcray** [comment]
> The point-of-instantiation can include tertiary methods in addition to or instead of the ones in the module.

Can it?  I was thinking it was only primary and secondary (and that that's why defining an `int.hash()` in k-nucleotide.chpl didn't cause it to get called).  But this test seems to show you're right:

```chapel
module MC {
  record C {
    type t;
    var x: t;
    proc dostuff() {
      x.foo();
      x.bar();
      x.baz();
    }
  }

}

module MR {
  record R {
    proc foo() { writeln("In R.foo"); }
  }

  proc R.bar() { writeln("In R.bar"); }
}

module MRT {
  use MR;
  proc R.baz() { writeln("In R.baz"); }
}

module Main {
  proc main() {
    use MC;
    use MR;
    use MRT;
    var myC: C(R);
    myC.dostuff();
  }
}
```

So then why didn't defining `int.hash()` in knucleotide.chpl override the built-in hash?  Or did we jump to that conclusion mistakenly?  Ah, checking https://github.com/chapel-lang/chapel/issues/18291#issuecomment-912776238 it's because there was an `int.hash()` already visible to our HashTable implementation.  So in  my example above, if MC defined an R.baz(), or otherwise saw one, it would win over the tertiary method in MRT (verifying, that's true).

OK, I think I'm back on the right page again, though it's a good example of how challenging POI can be to internalize.

[As an aside, this makes me wonder whether we could structure our hash table implementation so that it wouldn't see any of the built-in `.hash()` methods directly and go them all through POI so that a user could override the hash for a given module scope.  Of course, that might require making `int` more of a record type as we've often discussed than it currently is as I'm not sure that `int` is subject to POI?]

**Comment by mppf** [comment]
Yeah. 

Regarding your aside -- I'm worried that having different hash functions in different scopes could be a big problem if the hashtable is passed around, say. It would be very bad for a single hashtable instance to have two different has functions used, depending on where it is called from. So, I think that hash functions should not be allowed to be tertiary methods. (Also `int` is not special for POI and you can make tertiary methods on `int` available through POI just like any record).

**Comment by bradcray** [comment]
OK, so now that I'm back on the right path, returning to your original comment:

> One issue is that I don't think we'd want a Reflection query to find a tertiary method for one of these. Rather, if it does not exist at the record/class definition module, it should be defined by the compiler / use the default implementation.

Can you say more about why you think this, if the default implementation is defined in a module?  In past conversations, when we've focused on the compiler introducing the routine, I've agreed, thinking of the rationale as "because the compiler shouldn't have to worry about whether _any_ module defines the routine as a tertiary method and should make the decision based on information local to the module defining the type [for simplicity / tractability]."  But if we move the logic out of the compiler and into a module (such that no real analysis is needed beyond that which is required to do the reflection and call the appropriate routine), what would be the downside of having a tertiary method provide the implementation?  (e.g., "typically R's print/hash in thus-and-such a way, but I can override it in  my program by adding this additional module which provides a tertiary method.")  It's not immediately obvious to me.

Oh wait, maybe you answered the question in your comment that just showed up:  That the problem is that a given instance of a data type would behave differently if printed/hashed from a different source location.  Is that right?

**Comment by mppf** [comment]
> Oh wait, maybe you answered the question in your comment that just showed up: That the problem is that a given instance of a data type would behave differently if printed/hashed from a different source location. Is that right?

Exactly. You brought this up in a comment a long time ago when we were talking about custom .hash functions.

**Comment by bradcray** [comment]
My brain is obviously full and I need a new one.

I'm going to hide all of the responses between your original point about tertiary methods and this one so that they're still here for anyone who shares my confusion / lack of medium-term memory, but don't bog down the main thread of conversation.  Thanks!
